Tagging API Calls

aklarfeld opened the pull request:
Tags are able to be passed into the initialization function, or within a context, and then applied to a set of API calls.

There are currently two ways to tag API calls:

1. Simply pass the tags on initialization, e.g.

```
   Supergood.init(
      {
        config: { ...SUPERGOOD_CONFIG, allowLocalUrls: true },
        clientId: SUPERGOOD_CLIENT_ID,
        clientSecret: SUPERGOOD_CLIENT_SECRET,
        tags: {
          customTag: 'customValue'
        }
      },
      SUPERGOOD_SERVER
    );
```

2. Open up a context for specific API calls, to tag:
```
      Supergood.withTags({ userId: '123-34-5678' }, async () => {
        const balanceResponse = await client.accountsBalanceGet({
          access_token: ACCESS_TOKEN,
        });
        prettyPrintResponse(balanceResponse);
        response.json(balanceResponse.data);
      })
```

josiah-roberts commented:
What happens if we enter a Supergood context multiple times?
```typescript
      Supergood.init({ tags: { workload: 'backend' } });
      
      Supergood.withContext({ trip: 'in-n-out' }, async () => {
        await Supergood.withContext({ task: 'drive-there' }, async () => {
          await driveThereHttp();
        });
        await Supergood.withContext({ task: 'order' }, async () => {
          await orderHttp();
        });
      })
   ```
 
 I think you'd want to merge, but you may have to manage that behavior specifically.

aklarfeld commented:
> What happens if we enter a Supergood context multiple times?
> 
> ```ts
>       Supergood.init({ tags: { workload: 'backend' } });
>       
>       Supergood.withContext({ trip: 'in-n-out' }, async () => {
>         await Supergood.withContext({ task: 'drive-there' }, async () => {
>           await driveThereHttp();
>         });
>         await Supergood.withContext({ task: 'order' }, async () => {
>           await orderHttp();
>         });
>       })
> ```
> 
> I think you'd want to merge, but you may have to manage that behavior specifically.

Good call. I just fixed this and added some tests.

josiah-roberts commented:
One other nit; `withContext` might be `withTags`. Just to avoid creating more names of things, given it's called `tags` in config.

aklarfeld commented:
> One other nit; `withContext` might be `withTags`. Just to avoid creating more names of things, given it's called `tags` in config.

Yeah it's a fair nit, fixing now!
